jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,369 @@
+            Entry entry = new Entry(request, response, callback, priority);
+            queues.get(priority).offer(entry);
+            timeouts.schedule(entry);
+            return true;
+        }
+    }
+
+    /**
+     * <p>Returns the priority of the given suspended request.</p>
+     *
+     * @param request the suspended request to compute the priority for
+     * @return the priority of the given suspended request
+     * @see #setMaxPriority(int)
+     */
+    protected int getPriority(Request request)
+    {
+        return 0;
+    }
+
+    private void processAndComplete(Request request, Response response, Callback callback)
+    {
+        try
+        {
+            boolean handled = process(request, response, callback);
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} handled={} {}", this, handled, request);
+            if (!handled)
+                Response.writeError(request, response, callback, HttpStatus.NOT_FOUND_404);
+        }
+        catch (Throwable x)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} failed {}", this, request, x);
+            Response.writeError(request, response, callback, x);
+        }
+    }
+
+    private void resume()
+    {
+        semaphore.release();
+        ListIterator<Queue<Entry>> iterator = queues.listIterator(queues.size());
+        while (iterator.hasPrevious())
+        {
+            Queue<Entry> queue = iterator.previous();
+            Entry entry = queue.poll();
+            if (entry != null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("{} resuming {}", this, entry.request);
+                // Try again to acquire a semaphore pass.
+                // If it cannot be acquired, the request may
+                // be re-prioritized with a higher priority.
+                processAndComplete(entry.request, entry.response, entry.callback);
+                return;
+            }
+        }
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} no suspended requests to resume", this);
+    }
+
+    private class Entry implements CyclicTimeouts.Expirable
+    {
+        private final Request request;
+        private final Response response;
+        private final Callback callback;
+        private final int priority;
+        private final long expireNanoTime;
+
+        private Entry(Request request, Response response, Callback callback, int priority)
+        {
+            this.request = request;
+            this.response = response;
+            this.callback = callback;
+            this.priority = priority;
+            Duration maxSuspend = getMaxSuspend();
+            this.expireNanoTime = maxSuspend.isZero() ? Long.MAX_VALUE : NanoTime.now() + maxSuspend.toNanos();
+        }
+
+        @Override
+        public long getExpireNanoTime()
+        {
+            return expireNanoTime;
+        }
+
+        private void expire()
```

Comment:
Can this be delegated to a protected method on the handler so expiration actions can be extended?   Or is there not much they can do other than write and error?  I guess they could do something like dispatch handling to a low priority thread pool?